src/allmydata/testing/web.py

```diff
@@ -0,0 +1,286 @@
+    getattr(allmydata.uri, t).BASE_STRING
+    for t in dir(allmydata.uri)
+    if hasattr(getattr(allmydata.uri, t), 'BASE_STRING')
+]
+
+
+def capability_generator(kind):
+    """
+    Deterministically generates a streap of valid capabilities of the
+    given kind. The N, K and size values aren't related to anything
+    real.
+
+    :param str kind: the kind of capability, like `URI:CHK`
+
+    :returns: a generator that yields new capablities of a particular
+        kind.
+    """
+    if kind not in KNOWN_CAPABILITIES:
+        raise ValueError(
+            "Unknown capability kind '{} (valid are {})'".format(
+                kind,
+                ", ".join(KNOWN_CAPABILITIES),
+            )
+        )
+    # what we do here is to start with empty hashers for the key and
+    # ueb_hash and repeatedly feed() them a zero byte on each
+    # iteration .. so the same sequence of capabilities will always be
+    # produced. We could add a seed= argument if we wanted to produce
+    # different sequences.
+    number = 0
+    key_hasher = hashlib.new("sha256")
+    ueb_hasher = hashlib.new("sha256")  # ueb means "URI Extension Block"
+
+    # capabilities are "prefix:<128-bits-base32>:<256-bits-base32>:N:K:size"
+    while True:
+        number += 1
+        key_hasher.update("\x00")
+        ueb_hasher.update("\x00")
+
+        key = base32.b2a(key_hasher.digest()[:16])  # key is 16 bytes
+        ueb_hash = base32.b2a(ueb_hasher.digest())  # ueb hash is 32 bytes
+
+        cap = u"{kind}{key}:{ueb_hash}:{n}:{k}:{size}".format(
+            kind=kind,
+            key=key,
+            ueb_hash=ueb_hash,
+            n=1,
+            k=1,
+            size=number * 1000,
+        )
+        yield cap.encode("ascii")
+
+
+@attr.s
+class _FakeTahoeUriHandler(Resource, object):
+    """
+    An in-memory fake of (some of) the `/uri` endpoint of a Tahoe
+    WebUI
+    """
+
+    isLeaf = True
+
+    data = attr.ib(default=attr.Factory(dict))
+    capability_generators = attr.ib(default=attr.Factory(dict))
+
+    def _generate_capability(self, kind):
+        """
+        :param str kind: any valid capability-string type
+
+        :returns: the next capability-string for the given kind
+        """
+        if kind not in self.capability_generators:
+            self.capability_generators[kind] = capability_generator(kind)
+        capability = next(self.capability_generators[kind])
+        return capability
+
+    def add_data(self, kind, data):
+        """
+        adds some data to our grid
+
+        :returns: a two-tuple: a bool (True if the data is freshly added) and a capability-string
+        """
+        if not isinstance(data, bytes):
+            raise TypeError("'data' must be bytes")
+
+        for k in self.data:
+            if self.data[k] == data:
+                return (False, k)
+
+        cap = self._generate_capability(kind)
+        assert cap not in self.data, "logic error"  # shouldn't happen
```

Comment:
Probably should be some other kind of check.  Raise a `ValueError` maybe?  Or derive a new exception type and raise that.